src/main/java/org/pac4j/vertx/handler/impl/RequiresAuthenticationHandler.java

```diff
@@ -39,73 +46,130 @@
-    } else {
-      final VertxWebContext webContext = new VertxWebContext(routingContext);
-      final Client client = findClient(webContext, this.clientName, allowDynamicClientSelection);
-      final ProfileManager profileManager = new ProfileManager(webContext);
-      final AuthenticationFlow authenticationFlow = getAuthenticationFlow(client);
-
-      UserProfile profile = profileManager.get(authenticationFlow.useSession(webContext));
-      if (profile != null) {
-        // We have been authenticated, are we authorized? Use vert.x subsystem for this
-        authorise(new Pac4jUser(profile), routingContext);
-      } else {
-        // We have not been authenticated yet, so start of the authentication process
-        authenticationFlow.initiate(webContext, new Consumer<UserProfile>() {
-          @Override
-          public void accept(UserProfile userProfile) {
-            final Pac4jUser pac4jUser = new Pac4jUser(userProfile);
-            if (userProfile != null) {
-              profileManager.save(authenticationFlow.useSession(webContext), userProfile);
-              // Writing into the routing context now means that if the UserSessionHandler is in use then nothing
-              // else needs to be done to retrieve the user from the session and write to the routing context
-              routingContext.setUser(pac4jUser);
+    private static final Logger LOG = LoggerFactory.getLogger(RequiresAuthenticationHandler.class);
+
+    protected final Config config;
+    protected final String clientName;
+    protected final Vertx vertx;
+
+    protected HttpActionAdapter httpActionAdapter = new DefaultHttpActionAdapter();
+    protected ClientFinder clientFinder = new DefaultClientFinder();
+
+    public RequiresAuthenticationHandler(final Vertx vertx, final Config config, final Pac4jAuthProvider authProvider,
+                                         final Pac4jAuthHandlerOptions options) {
+        super(authProvider);
+        clientName = options.clientName();
+        this.vertx = vertx;
+        this.config = config;
+    }
+
+    // Port of Pac4J auth to a handler in vert.x 3.
+    @Override
+    public void handle(RoutingContext routingContext) {
+
+        final User user = routingContext.user();
+        if (user != null) {
+            // Already logged in, just authorise
+            authorise(user, routingContext);
+        } else {
+            final VertxWebContext webContext = new VertxWebContext(routingContext);
+            final List<Client> currentClients = clientFinder.find(config.getClients(), webContext, this.clientName);
+            final ProfileManager profileManager = new ProfileManager(webContext);
+
+            UserProfile profile = profileManager.get(useSession(webContext, currentClients));
+            if (profile != null) {
+                // We have been authenticated, are we authorized? Use vert.x subsystem for this
+                authorise(new Pac4jUser(profile), routingContext);
```

Comment:
By the way, it's not nitpicking. I considered throwing away the vert.x user altogether but then ran into use of the UserSessionHandler (which I wasn't previously aware of) on the vert.x group, so thought we should support this properly once I'd found out about it. If I hadn't actually found out about it, I'd have stripped this code out :)